jupyterlab_gcedetails/src/service/price_service.ts

```diff
@@ -0,0 +1,219 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ServerConnection } from '@jupyterlab/services';
+import { Option, HardwareConfiguration } from '../data/data';
+
+/** The monthly price catalog for region and sku */
+type Catalog = Map<string, Map<string, number>>;
+
+/** The average number of hours in a month, i.e., 365 / 12 * 24 */
+const HOURS_PER_MONTH = 730;
+const NO_SUSTAINED_DISCOUNT_PREFIXES = ['e2-'];
+const REGIONS: Option[] = [
+  {
+    value: 'us-central1',
+    text: 'us-central1 (Iowa)',
+  },
+  {
+    value: 'us-east1',
+    text: 'us-east1 (South Carolina)',
+  },
+  {
+    value: 'us-east4',
+    text: 'us-east4 (Northern Virginia)',
+  },
+  {
+    value: 'us-west1',
+    text: 'us-west1 (Oregon)',
+  },
+  {
+    value: 'us-west2',
+    text: 'us-west2 (Los Angeles)',
+  },
+  {
+    value: 'us-west3',
+    text: 'us-west3 (Salt Lake City)',
+  },
+  {
+    value: 'asia-east1',
+    text: 'asia-east1 (Taiwan)',
+  },
+  {
+    value: 'europe-north1',
+    text: 'europe-north1 (Finland)',
+  },
+  {
+    value: 'europe-west1',
+    text: 'europe-west1 (Belgium)',
+  },
+  {
+    value: 'europe-west2',
+    text: 'europe-west2 (London)',
+  },
+  {
+    value: 'europe-west3',
+    text: 'europe-west3 (Frankfurt)',
+  },
+  {
+    value: 'europe-west4',
+    text: 'europe-west4 (Netherlands)',
+  },
+  {
+    value: 'europe-west6',
+    text: 'europe-west6 (Zurich)',
+  },
+  {
+    value: 'asia-east1',
+    text: 'asia-east1 (Taiwan)',
+  },
+  {
+    value: 'asia-east2',
+    text: 'asia-east2 (Hong Kong)',
+  },
+  {
+    value: 'asia-south1',
+    text: 'asia-south1 (Mumbai)',
+  },
+  {
+    value: 'asia-northeast1',
+    text: 'asia-northeast1 (Tokyo)',
+  },
+  {
+    value: 'asia-northeast2',
+    text: 'asia-northeast2 (Osaka)',
+  },
+  {
+    value: 'asia-northeast3',
+    text: 'asia-northeast3 (Seoul)',
+  },
+  {
+    value: 'asia-southeast1',
+    text: 'asia-southeast1 (Singapore)',
+  },
+];
+
+export class PriceService {
+  private catalog: Catalog;
+  private _receivedError: boolean;
+
+  async getPriceList(): Promise<void> {
```

Comment:
This method is a bit of misnomer since it really doesn't return anything like the prefix `get` would entail. I'd suggest renaming it to something like loadPrices, or perhaps consider implementing the lazy-loading strategy from some of the other async methods where the first call retrieves the content from the remote API and the subsequent calls return the same promise.

Reply:
Great, I'll update the name but I believe the way I have implemented does use cached data as it only runs the method if the catalog hasn't yet been built or was built unsuccessfully. Is this sufficient or would it still make sense to implement the lazy-loading strategy as you suggested?

Reply:
My suggestion was more about not making the `getPriceList` method the public interface from other components since there's really never a case where the entire pricing is shown, but rather the lookup is always based on a particular configuration. In doing this, you could hide the logic that actually retrieves the prices from any other component and make it easier to swap out in the future in the case where we might actually have a dedicated GCP service for returning an exact price based on a customer's project and resource configurations.